In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [4]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!

# Review the DataFrame
# YOUR CODE HERE!

file_path = '/Users/zan/Desktop/Data Analytics Course/credit-risk-classification/Credit_Risk/Resources/lending_data.csv'
df = pd.read_csv(file_path)


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [5]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']



# Separate the X variable, the features
X = df.drop(columns=['loan_status'])

In [6]:
# Review the y variable Series
print("First few rows of y:")
print(y.head())

print("\nUnique values in y:")
print(y.unique())

print("\nValue counts for y:")
print(y.value_counts())

print("\nData type of y:")
print(y.dtype)

print("\nNumber of missing values in y:")
print(y.isnull().sum())


First few rows of y:
0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

Unique values in y:
[0 1]

Value counts for y:
loan_status
0    75036
1     2500
Name: count, dtype: int64

Data type of y:
int64

Number of missing values in y:
0


In [7]:
# Review the X variable DataFrame
print("First few rows of X:")
print(X.head())

print("\nInfo about X:")
print(X.info())

print("\nStatistical summary of X:")
print(X.describe())

print("\nNumber of missing values in each column of X:")
print(X.isnull().sum())

print("\nShape of X:")
print(X.shape)

print("\nColumns in X:")
print(X.columns)


First few rows of X:
   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  

Info about X:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     7

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)


Shape of X_train: (62028, 7)
Shape of X_test: (15508, 7)
Shape of y_train: (62028,)
Shape of y_test: (15508,)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

print("Logistic Regression model has been successfully trained.")


Logistic Regression model has been successfully trained.


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
y_pred = model.predict(X_test)

# Display the first few predictions
print("First few predictions:")
print(y_pred[:10])

# Compare with the actual labels
print("\nFirst few actual labels:")
print(y_test[:10].values)


First few predictions:
[0 0 0 0 0 0 0 0 0 0]

First few actual labels:
[0 0 0 0 0 0 0 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Generate a confusion matrix for the model
cm = confusion_matrix(y_test, y_pred)

# Display the confusion matrix
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[14924    77]
 [   31   476]]


In [12]:
# Print the classification report for the model
report = classification_report(y_test, y_pred)

# Print the classification report
print("Classification Report:")
print(report)


Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.86      0.94      0.90       507

    accuracy                           0.99     15508
   macro avg       0.93      0.97      0.95     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The model identifies healthy loans 99% of the time and all the loans identified as healthy were healthy. Although the model demonstrated excellent precision in identifying healthy loans, 86% of the loans identified high risk by the model were correct. Despite the lower precision and recall, the model still had a good f1-score of 90%. Overall, the model has a good balance in identifying healthy loans, it is recommended to use the model.

---